--- src.orig/imx_ext.c	2012-10-01 02:18:44.005430995 +0200
+++ src/imx_ext.c	2012-10-01 02:19:33.575431373 +0200
@@ -90,10 +90,10 @@
 	/* Check if any reply values need byte swapping */
 	if (client->swapped)
 	{
-		swaps(&rep.sequenceNumber, n);
-		swapl(&rep.length, n);
-		swapl(&rep.pixmapPhysAddr, n);
-		swapl(&rep.pixmapPitch, n);
+		swaps(&rep.sequenceNumber);
+		swapl(&rep.length);
+		swapl(&rep.pixmapPhysAddr);
+		swapl(&rep.pixmapPitch);
 	}
 
 	/* Reply to client */
@@ -121,10 +121,10 @@
 
 	REQUEST(xIMX_EXT_GetPixmapPhysAddrReq);
 
-	swaps(&stuff->length, n);
+	swaps(&stuff->length);
 	REQUEST_SIZE_MATCH(xIMX_EXT_GetPixmapPhysAddrReq);
 
-	swapl(&stuff->pixmap, n);
+	swapl(&stuff->pixmap);
 	return Proc_IMX_EXT_GetPixmapPhysAddr(client);
 }
 
